Libraries/Blob/Blob.js

```diff
@@ -0,0 +1,156 @@
+ *
+ * The API is modeled after the W3C Blob API, with one caveat
+ * regarding explicit deallocation. Refer to the `close()`
+ * method for further details.
+ *
+ * Example usage in a React component:
+ *
+ *   class WebSocketImage extends React.Component {
+ *      state = {blob: null};
+ *      componentDidMount() {
+ *        let ws = this.ws = new WebSocket(...);
+ *        ws.binaryType = 'blob';
+ *        ws.onmessage = (event) => {
+ *          if (this.state.blob) {
+ *            this.state.blob.close();
+ *          }
+ *          this.setState({blob: event.data});
+ *        };
+ *      }
+ *      componentUnmount() {
+ *        if (this.state.blob) {
+ *          this.state.blob.close();
+ *        }
+ *        this.ws.close();
+ *      }
+ *      render() {
+ *        if (!this.state.blob) {
+ *          return <View />;
+ *        }
+ *        return <Image source={{uri: URL.createObjectURL(this.state.blob)}} />;
+ *      }
+ *   }
+ */
+class Blob {
+  /**
+   * Size of the data contained in the Blob object, in bytes.
+   */
+  size: number;
+  /*
+   * String indicating the MIME type of the data contained in the Blob.
+   * If the type is unknown, this string is empty.
+   */
+  type: string;
+
+  /*
+   * Unique id to identify the blob on native side (non-standard)
+   */
+  blobId: string;
+  /*
+   * Offset to indicate part of blob, used when sliced (non-standard)
+   */
+  offset: number;
+
+  /**
+   * Construct blob instance from blob data from native.
+   * Used internally by modules like XHR, WebSocket, etc.
+   */
+  static create(props: BlobProps): Blob {
+    return Object.assign(Object.create(Blob.prototype), props);
+  }
+
+  /**
+   * Constructor for JS consumers.
+   * Currently we only support creating Blobs from other Blobs.
+   */
+  constructor(parts: Array<Blob>, options: any) {
+    let blobId = createUUID();
+    let size = 0;
+    parts.forEach((part) => {
+      if (!(part instanceof Blob)) {
+        throw new Error('Can currently only create a Blob from other Blobs');
+      }
+      size += part.size;
+    });
+    BlobModule.createFromParts(parts, blobId);
+    return Blob.create({
+      blobId,
+      offset: 0,
+      size,
+    });
+  }
+
+  slice(start?: number, end?: number): Blob {
+    let offset = this.offset;
+    let size = this.size;
+    if (typeof start === 'number') {
+      if (start > size) {
+        start = size;
+      }
+      offset += start;
+      size -= start;
+
+      if (typeof end === 'number') {
+        if (end < 0) {
+          end = this.size + end;
+        }
+        size = end - start;
+      }
+    }
+    return Blob.create({
+      blobId: this.blobId,
+      offset,
+      size,
+    });
+  }
+
+  /**
+   * This method is in the standard, but not actually implemented by
+   * any browsers at this point. It's important for how Blobs work in
+   * React Native, however, since we cannot de-allocate resources automatically,
```

Comment:
You could hook into JSC's finalizer callback to clean this up, but we currently don't have a good cross-platform approach to creating custom JS objects. It also would be an issue with the Websocket executor.